### PR TITLE
[float8] emulate eager mode precision casts for float8 rowwise

### PR DESCRIPTION
**Pull request by danielvegamyhre**
## Summary
This is a short term solution for #1056 (pytorch core issue tracker: https://github.com/pytorch/pytorch/issues/150859)

For a long term solution, the compiler folks are looking into supporting deterministic recompute of intermediates for AC during the backward pass.

## Test plan
- Manual testing confirmed loss decreases as expected without NaNs (training logs: https://www.internalfb.com/phabricator/paste/view/P1787321079

cc @tianyu-l @lessw2020

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
fyi @drisspg @ngimel @bdhirsh

**Review by lessw2020 — APPROVED** [review]
great job tracing this thing down and temp fix here  == lgtm!

**Comment by tianyu-l** [comment]
We should do it in `float8.py`, rather than `train.py`.
@danielvegamyhre

**Comment by danielvegamyhre** [comment]
> We should do it in `float8.py`, rather than `train.py`. @danielvegamyhre

Sure, moved it here https://github.com/pytorch/torchtitan/pull/1110. Sorry for the premature merge.
